feat(syncthing, fs): allow setting tmpdir on command line (fixes #2208)

stuartnelson3 opened the pull request:
### Purpose

Allow users to specify a tempdir where they want to store all files as they're being downloaded.

Requested in https://github.com/syncthing/syncthing/issues/2208

### Testing

Looking for help with manual and automated testing. So far, I've only verified that the project compiles and print statements are being executed.

### Documentation

https://github.com/syncthing/docs/pull/923

## Authorship

Your name and email will be added automatically to the AUTHORS file
based on the commit metadata.

stuartnelson3 commented:
Force push was cleaning up my commit messages

marbens-arch commented:
> Force push was cleaning up my commit messages

The PR title is the commit message BTW, as PRs are squashed and merged. The style of commits within your branch doesn't matter.

marbens-arch commented on man/syncthing.1:
```diff
@@ -38,7 +38,7 @@ syncthing \- Syncthing
           [\-\-audit] [\-\-auditfile=<file|\-|\-\->] [\-\-browser\-only] [\-\-device\-id]
           [\-\-generate=<dir>] [\-\-gui\-address=<address>] [\-\-gui\-apikey=<key>]
-          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>]
+          [\-\-home=<dir> | \-\-config=<dir> \-\-data=<dir>] [\-\-tmp=<tmp>]
```
The manpages are auto-generated from the docs repository, and shouldn't be changed directly. Please edit [users/syncthing.rst](https://github.com/syncthing/docs/blob/main/users/syncthing.rst?plain=1) in docs instead.

marbens-arch commented on AUTHORS:
```diff
@@ -323,6 +323,7 @@ Sonu Kumar Saw <31889738+dev-saw99@users.noreply.github.com>
 Stefan Kuntz (Stefan-Code) <stefan.github@gmail.com> <Stefan.github@gmail.com>
 Stefan Tatschner (rumpelsepp) <stefan@sevenbyte.org> <rumpelsepp@sevenbyte.org> <stefan@rumpelsepp.org>
 Steven Eckhoff <steven.eckhoff.opensource@gmail.com>
+Stuart Nelson <hi@stuartnelson.xyz>
```
This is not necessary, as you will be automatically added here. Not sure if it's bad practice to add it manually though.

stuartnelson3 replied:
Ah, understood. I misunderstood the contributing guide: https://docs.syncthing.net/dev/contributing.html#tests

tomasz1986 commented:
Thank you for working on this!

I would just like to ask a few questions. Please keep in mind that I have not actually tested the PR yet.

1. What is going to happen if multiple temporary files have the exact same names?
2. Can the temp folder be located on a different drive/partition? If yes, what is going to happen after the transfer has finished? Normally, the original file is deleted and replaced by the new one. However, with the new file located on different storage, it may take a while to move it to the original location. Also, what if the temporary file is too large and doesn't fit on the main storage? Similarly, what if there is not enough space on the temporary storage?
3. Unfinished downloads are normally removed after 24 hours (by default) during a folder rescan. Is this going to work with them located in the temp folder as well?

tomasz1986 commented on cmd/syncthing/main.go:
```diff
@@ -144,6 +144,7 @@ type serveOptions struct {
 	AuditFile        string `name:"auditfile" placeholder:"PATH" help:"Specify audit file (use \"-\" for stdout, \"--\" for stderr)"`
 	BrowserOnly      bool   `help:"Open GUI in browser"`
 	DataDir          string `name:"data" placeholder:"PATH" env:"STDATADIR" help:"Set data directory (database and logs)"`
+	TempDir          string `name:"tmp" placeholder:"TMPDIR" env:"STTMPDIR" help:"Set single tempdir for storing incomplete files (optional)"`
```
You've got a mix of "temp" and "tmp" in many different places. Maybe streamline everything to just "temp" (and "TEMPDIR", "STTEMPDIR", "tempdir", etc.)?

stuartnelson3 replied:
Sure, not an issue

stuartnelson3 commented:
> Thank you for working on this!

My pleasure!

> I would just like to ask a few questions. Please keep in mind that I have not actually tested the PR yet.

Responses are inline:

>     1. What is going to happen if multiple temporary files have the exact same names?

The temp files have their full path hashed and use that for their temp name. This was mentioned in the issue, and I've added a test for it. I've also noted this in the documentation PR (https://github.com/syncthing/docs/pull/923)

>     2. Can the temp folder be located on a different drive/partition? If yes, what is going to happen after the transfer has finished? Normally, the original file is deleted and replaced by the new one. However, with the new file located on different storage, it may take a while to move it to the original location. Also, what if the temporary file is too large and doesn't fit on the main storage? Similarly, what if there is not enough space on the temporary storage?

>Can the temp folder be located on a different drive/partition? If yes, what is going to happen after the transfer has finished?

The user is able to place the temp folder on a different drive/partition, but the rename of the tempfile after the download finishes will fail.

>Normally, the original file is deleted and replaced by the new one. However, with the new file located on different storage, it may take a while to move it to the original location.

Ah! Can you point me to where this happens in the code? From reading the issue, it sounded like it was a simple rename. I implemented this PR based off this comment in the issue: https://github.com/syncthing/syncthing/issues/2208#issuecomment-549420078

>Also, what if the temporary file is too large and doesn't fit on the main storage? Similarly, what if there is not enough space on the temporary storage?

I'm new to the project, so I'm not sure how you decide to deal with these sorts of issues. I personally would consider this "user error", the same as if they were trying to sync to a drive that ran out of space, and consider it the user's responsibility to configure a temp dir that has sufficient space.

>     3. Unfinished downloads are normally removed after 24 hours (by default) during a folder rescan. Is this going to work with them located in the temp folder as well?

This is a feature I have no investigated. Can you point me to where in the code this happens, and I can try to see what the behavior will be?

tomasz1986 commented:
> > Can the temp folder be located on a different drive/partition? If yes, what is going to happen after the transfer has finished?
> 
> The user is able to place the temp folder on a different drive/partition, but the rename of the tempfile after the download finishes will fail.
> 
> > Also, what if the temporary file is too large and doesn't fit on the main storage? Similarly, what if there is not enough space on the temporary storage?
> 
> I'm new to the project, so I'm not sure how you decide to deal with these sorts of issues. I personally would consider this "user error", the same as if they were trying to sync to a drive that ran out of space, and consider it the user's responsibility to configure a temp dir that has sufficient space.

With these in mind, I think it would be good to have a check whether the temp location belongs to the same partition, and only then allow to use it. If not, there should be an error message about it. This would also solve all possible temp storage space-related issues 🙂.

> > 3. Unfinished downloads are normally removed after 24 hours (by default) during a folder rescan. Is this going to work with them located in the temp folder as well?
> 
> This is a feature I have no investigated. Can you point me to where in the code this happens, and I can try to see what the behavior will be?

It is related to

https://github.com/search?q=repo%3Asyncthing%2Fsyncthing%20IsTemporary&type=code

and also

https://github.com/search?q=repo%3Asyncthing%2Fsyncthing%20TempLifetime&type=code

bt90 commented:
> With these in mind, I think it would be good to have a check whether the temp location belongs to the same partition

Good luck with that :sweat_smile:

calmh commented:
I think this may make more sense as an advanced folder setting, so that it is settable per folder?

stuartnelson3 commented:
> I think this may make more sense as an advanced folder setting, so that it is settable per folder?

Would that be done in the `FolderConfiguration` struct? https://github.com/syncthing/syncthing/blob/main/lib/config/folderconfiguration.go#L50

stuartnelson3 commented:
@tomasz1986 

> > > Can the temp folder be located on a different drive/partition? If yes, what is going to happen after the transfer has finished?
> > 
> > 
> > The user is able to place the temp folder on a different drive/partition, but the rename of the tempfile after the download finishes will fail.
> > > Also, what if the temporary file is too large and doesn't fit on the main storage? Similarly, what if there is not enough space on the temporary storage?
> > 
> > 
> > I'm new to the project, so I'm not sure how you decide to deal with these sorts of issues. I personally would consider this "user error", the same as if they were trying to sync to a drive that ran out of space, and consider it the user's responsibility to configure a temp dir that has sufficient space.
> 
> With these in mind, I think it would be good to have a check whether the temp location belongs to the same partition, and only then allow to use it. If not, there should be an error message about it. This would also solve all possible temp storage space-related issues 🙂.
> 
> > > 3. Unfinished downloads are normally removed after 24 hours (by default) during a folder rescan. Is this going to work with them located in the temp folder as well?
> > 
> > 
> > This is a feature I have no investigated. Can you point me to where in the code this happens, and I can try to see what the behavior will be?
> 
> It is related to
> 
> https://github.com/search?q=repo%3Asyncthing%2Fsyncthing%20IsTemporary&type=code
> 
> and also
> 
> https://github.com/search?q=repo%3Asyncthing%2Fsyncthing%20TempLifetime&type=code

I looked at the `IsTemporary` code, and it should be fine. Even when hashing the file name to prevent conflict, the code still prepends the OS-specific prefix. This is how the code decides if a file is a temp file or not. I added a test to confirm this.

stuartnelson3 commented:
I looked at the failing issues from my latest commit, it looks like what's flagging the failure are pre-existing issues in the code from the files I touched? Maybe I'm parsing the output incorrectly.

stuartnelson3 commented on lib/model/folder_sendrecv_test.go:
Running into some issues when I try to use a tempdir.

It seems the underlying fake FS only accepts `var tempdir = ".stfolder"`, otherwise the file creation is rejected.

There are some other failures that occur, but I'm not sure how deep to look into the internals. It might be some simple test setup issue or a deeper problem.

For example:

```
# go test ./lib/model/... -run TestHandleFileWithTemp -v
=== RUN   TestHandleFileWithTemp
2025/05/06 13:15:12 INFO: No stored folder metadata for "default"; recalculating
2025/05/06 13:15:12 INFO: Ready to synchronize "default" (default) (sendreceive)
2025/05/06 13:15:12 INFO: Completed initial scan of sendreceive folder "default" (default)
    folder_sendrecv_test.go:207: Unexpected count of copy blocks: 8 != 4
--- FAIL: TestHandleFileWithTemp (0.01s)
FAIL
FAIL	github.com/syncthing/syncthing/lib/model	0.527s
?   	github.com/syncthing/syncthing/lib/model/mocks	[no test files]
FAIL
```

rasa commented:
> I think it would be good to have a check whether the temp location belongs to the same partition, and only then allow to use it.

I just published a package, [compat](https://github.com/rasa/compat), that provides this functionality:
```go
package main

import (
    "fmt"
    "os"
    "github.com/rasa/compat"
)

func main() {
   file1 := os.Argv[1]
   file2 := os.Argv[2]
   fmt.Printf("SameDevices(%q, %q)=%v\n", file1, file2, compat.SameDevices(file1, file2))
}
```
Since it uses the same logic that go uses on [unix](https://github.com/golang/go/blob/c57b0d6c7e54284c8967530153774dbc62190b00/src/os/types_unix.go#L28), [Windows](https://github.com/golang/go/blob/c57b0d6c7e54284c8967530153774dbc62190b00/src/os/types_windows.go#L353), and [Plan9](https://github.com/golang/go/blob/c57b0d6c7e54284c8967530153774dbc62190b00/src/os/types_plan9.go#L26), it should work as expected, unless some file systems have non-unique device IDs, but my guess is that's pretty rare, if ever.

See https://pkg.go.dev/github.com/rasa/compat#SameDevices
